JetStreamCompose/jetstream/src/main/java/com/google/jetstream/data/repositories/MovieRepositoryImpl.kt

```diff
@@ -33,123 +34,141 @@ import com.google.jetstream.data.util.StringConstants.Movie.Reviewer.DefaultCoun
 import com.google.jetstream.data.util.StringConstants.Movie.Reviewer.DefaultRating
 import com.google.jetstream.data.util.StringConstants.Movie.Reviewer.FreshTomatoes
 import com.google.jetstream.data.util.StringConstants.Movie.Reviewer.ReviewerName
+import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.flow
-import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
 import kotlinx.serialization.json.Json
 import javax.inject.Inject
 
-class MovieRepositoryImpl @Inject constructor(assetsReader: AssetsReader) : MovieRepository {
+class MovieRepositoryImpl @Inject constructor(
+    private val assetsReader: AssetsReader
+) : MovieRepository {
 
-    private val mostPopularMovies: List<MoviesResponseItem> =
-        assetsReader.getJsonDataFromAsset(StringConstants.Assets.MostPopularMovies).map {
-            Json.decodeFromString<List<MoviesResponseItem>>(it)
-        }.getOrDefault(emptyList())
+    private val movieCastDataSource = CachedDataSource {
```

Comment:
I have two issues with those datasources:
1. Dependencies should be injected from outside the class not constructed internall
2. NIT: There's too many data sources here, I wish  it could be simplified somehow

Reply:
Thanks. Totally makes sense. Created four data source classes, MovieDataSource, TvDataSource, MovieCastDataSource, and MoviecategoryDataSource, which are injected from outside the class.